File: leetcode/C++/word-pattern.cpp
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        unordered_map<string, string> um, rum;
        unordered_map<string, string>::iterator it;
        stringstream ss(s);
        string temp, t;
        int idx = 0;
        while (getline(ss, temp, ' ')) {
            if(idx >= pattern.length()) return false;
            t = to_string(pattern[idx]);
            it = um.find(t);
            if(it == um.end()) um[t] = temp;
            else if(it->second != temp) return false;
            
            it = rum.find(temp);
            if(it == rum.end()) rum[temp] = t;
            else if(it->second != t) return false;
            
            ++idx;
        }
        return idx == pattern.length();
    }
};

File: leetcode/Python/word-pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        um, rum = dict(), dict()
        ps = s.split(' ')
        if len(ps) != len(pattern):
            return False
        for i in range(len(ps)):
            v = um.get(pattern[i], None)
            rv = rum.get(ps[i], None)
            if v is None:
                v = um[pattern[i]] = ps[i]
            if rv is None:
                rv = rum[ps[i]] = pattern[i]
            if v != ps[i] or rv != pattern[i]:
                return False
        return True